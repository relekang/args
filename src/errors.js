// @flow
type CliErrorOptions = {
  message: string,
  exitCode: number,
  showHelp?: boolean
};
export class CliError extends Error {
  exitCode: number;
  showHelp: boolean;

  constructor({ message, exitCode, showHelp }: CliErrorOptions) {
    super(message);
    this.exitCode = exitCode;
    this.showHelp = showHelp || false;
  }

  toString() {
    return this.message;
  }
}
