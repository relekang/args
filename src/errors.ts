type CliErrorOptions = {
  message: string;
  exitCode: number;
  showHelp?: boolean;
};
export class CliError extends Error {
  exitCode: number;
  showHelp: boolean;

  constructor({ message, exitCode, showHelp }: CliErrorOptions) {
    super(message);
    this.exitCode = exitCode;
    this.showHelp = showHelp || false;

    Object.setPrototypeOf(this, CliError.prototype);
  }

  toString() {
    return this.message;
  }
}
