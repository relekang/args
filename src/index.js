// @flow

export type Config = {};

export default function args(_config: Config) {
  return ([_node, _program, ..._rest]: Array<String>) => ({});
}
