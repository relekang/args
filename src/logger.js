// @flow
/* eslint-disable no-console */

export function error(...args: *) {
  console.error(...args);
}

export function log(...args: *) {
  console.log(...args);
}
