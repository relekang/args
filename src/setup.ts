import { CliError } from './errors';
import { Config } from './types';

export async function setup(config: Config) {
  if (config.needsSetup && (await config.needsSetup())) {
    if (config.setup) {
      await config.setup();
    } else {
      throw new CliError({
        message: 'This cli requires setup, however the setup is not provided.',
        exitCode: 1,
      });
    }
  }
}
