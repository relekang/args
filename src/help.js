// @flow
/* eslint-disable global-require */
import chalk from "chalk";
import wrap from "word-wrap";
import path from "path";
import fs from "fs";
import { promisify } from "util";

import type { Config, CommandConfig, CommandOption, Options } from "./types";

import * as logger from "./logger";

const readdir = promisify(fs.readdir);

function values<A>(o: { [key: string]: A }): Array<A> {
  // $FlowFixMe
  return Object.values(o);
}

async function getCommands(config): { [key: string]: CommandConfig } {
  if (config.commands) {
    return config.commands;
  }
  return (await readdir(config.commandsPath)).filter(/\.js$/.test).reduce(
    (lastValue, file) => ({
      ...lastValue,
      // $FlowFixMe
      [file.replace(/.js$/, "")]: require(path.join(config.commandsPath, file))
    }),
    {}
  );
}

export async function createCommandsList(config: Config) {
  const commands = await getCommands(config);

  const length = Math.max(
    0,
    ...values(commands).map(({ name }) => name.length)
  );
  return values(commands)
    .map(command => {
      const spacing = new Array(Math.max(1, length - command.name.length + 1))
        .fill(" ")
        .join("");
      return `${command.name}${spacing} - ${command.help || ""}`;
    })
    .join("\n");
}

export function createPositionalUsage(command: CommandConfig) {
  return (command.positionalOptions || [])
    .map(({ name, required }) => (required ? `<${name}>` : `[<${name}>]`))
    .join(" ");
}

export function createUsageString(name: string, command: CommandConfig) {
  const positionals = createPositionalUsage(command);
  return (
    `Usage: ${name} ${command.name}` +
    (positionals ? ` ${positionals}` : "") +
    (command.namedOptions ? " [--options]" : "")
  );
}

export function createOptionHelpString(option: CommandOption) {
  return `  - ${option.name}${option.help ? ` - ${option.help}` : ""}`;
}

export function createSubCommandHelp(name: string, command: CommandConfig) {
  let lines = [
    chalk`{bold ${command.name}} - ${command.help} `,
    "\n",
    createUsageString(name, command),
    command.manual ? "\n" + wrap(command.manual, { width: 80 }) : ""
  ];
  if (command.positionalOptions) {
    lines = [...lines, "", chalk.bold("Arguments:")];
    lines = [
      ...lines,
      ...(command.positionalOptions || []).map(createOptionHelpString)
    ];
  }
  if (command.namedOptions) {
    lines = [...lines, "", chalk.bold("Options:")];
    lines = [
      ...lines,
      ...(command.namedOptions || []).map(createOptionHelpString)
    ];
  }
  return lines.join("\n");
}

function findSubCommand(config: Config, name: string): ?CommandConfig {
  if (config.commands) {
    // $FlowFixMe
    return config.commands[name];
  } else {
    try {
      // $FlowFixMe
      return require(path.join(config.commandsPath, name));
    } catch (error) {
      return;
    }
  }
}

export async function help(config: Config, options: Options) {
  const dashes = "--------------------------";
  const subCommand = findSubCommand(config, options._[0]);
  logger.error(chalk`{gray ${dashes}} {bold ${config.name}} {gray ${dashes}}`);
  logger.error("");
  if (subCommand) {
    logger.error(createSubCommandHelp(config.name, subCommand));
  } else {
    logger.error(chalk.bold`Commands: `);
    logger.error(await createCommandsList(config));
  }
}
