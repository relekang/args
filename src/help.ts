/* eslint-disable global-require */
import chalk from 'chalk';
import wrap from 'word-wrap';
import path from 'path';
import fs from 'fs';
import { promisify } from 'util';

import { Config, CommandConfig, CommandOption, Options } from "./types";

import * as logger from './logger';

const readdir = promisify(fs.readdir);

function values<A>(o: { [key: string]: A }): Array<A> {
  // $FlowFixMe
  return Object.values(o);
}

async function getCommands(
  config: Config
): Promise<{ [key: string]: CommandConfig }> {
  // @ts-ignore
  if (config.commands) {
    // @ts-ignore
    return config.commands;
  }
  // @ts-ignore
  return (await readdir(config.commandsPath))
    .filter(file => /\.[jt]s$/.test(file))
    .reduce(
      (lastValue, file) => ({
        ...lastValue,
        [file.replace(/.[jt]s$/, "")]: require(path.join(
          // @ts-ignore
          config.commandsPath,
          file
        )),
      }),
      {}
    );
}

export async function createCommandsList(config: Config) {
  const commands = await getCommands(config);

  const length = Math.max(
    0,
    ...values(commands).map(({ name }) => name.length)
  );
  return values(commands)
    .map(command => {
      const spacing = new Array(Math.max(1, length - command.name.length + 1))
        .fill(' ')
        .join('');
      return `${command.name}${spacing} - ${command.help || ''}`;
    })
    .join('\n');
}

export function createPositionalUsage(command: CommandConfig) {
  return (command.positionalOptions || [])
    .map(({ name, required }) => (required ? `<${name}>` : `[<${name}>]`))
    .join(' ');
}

export function createUsageString(name: string, command: CommandConfig) {
  const positionals = createPositionalUsage(command);
  return (
    `Usage: ${name} ${command.name}` +
    (positionals ? ` ${positionals}` : '') +
    (command.namedOptions ? ' [--options]' : '')
  );
}

export function createOptionHelpString(option: CommandOption) {
  return `  - ${option.name}${option.help ? ` - ${option.help}` : ''}`;
}

export function createSubCommandHelp(name: string, command: CommandConfig) {
  let lines = [
    chalk`{bold ${command.name}} - ${command.help} `,
    '\n',
    createUsageString(name, command),
    command.manual ? '\n' + wrap(command.manual, { width: 80 }) : '',
  ];
  if (command.positionalOptions) {
    lines = [...lines, '', chalk.bold('Arguments:')];
    lines = [
      ...lines,
      ...(command.positionalOptions || []).map(createOptionHelpString),
    ];
  }
  if (command.namedOptions) {
    lines = [...lines, '', chalk.bold('Options:')];
    lines = [
      ...lines,
      ...(command.namedOptions || []).map(createOptionHelpString),
    ];
  }
  return lines.join('\n');
}

function findSubCommand(
  config: Config,
  name: string
): CommandConfig | undefined {
  // @ts-ignore
  if (config.commands) {
    // @ts-ignore
    return config.commands[name];
  } else {
    try {
      // @ts-ignore
      return require(path.join(config.commandsPath, name));
    } catch (error) {
      return undefined;
    }
  }
}

export async function help(config: Config, options: Options) {
  const dashes = '--------------------------';
  const subCommand = findSubCommand(config, options._[0]);
  logger.error(chalk`{gray ${dashes}} {bold ${config.name}} {gray ${dashes}}`);
  logger.error('');
  if (subCommand) {
    logger.error(createSubCommandHelp(config.name, subCommand));
  } else {
    logger.error(chalk.bold`Commands: `);
    logger.error(await createCommandsList(config));
  }
}
