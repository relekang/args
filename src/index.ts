/* eslint-disable global-require */
import mri from 'mri';
import chalk from 'chalk';

import { help } from './help';
import * as logger from './logger';
import { CliError } from './errors';
import { parse } from './parse';
import { setup } from './setup';
import { Config, Options } from './types';
import { findSubCommand } from './subCommand';

export * from './errors';

async function __args(config: Config, subCommand: string, args: Array<string>) {
  await setup(config);
  if (subCommand === 'help') {
    return help(config, mri(args));
  }

  const command = findSubCommand(config, subCommand);

  if (command) {
    const options: Options = parse(command, args);
    await command.run(options);
  } else {
    throw new CliError({
      message: 'Unknown command',
      exitCode: 1,
      showHelp: true,
    });
  }
}

export function args(config: Config) {
  return ([_node, _program, subCommand, ...rest]: Array<string>) => {
    return __args(config, subCommand || config.defaultCommand || 'help', rest)
      .catch(async error => {
        if (
          error.constructor === CliError ||
          error.constructor.constructor === CliError
        ) {
          if (error.showHelp) {
            logger.error('\n');
            logger.error(chalk.bold(chalk.red(error.toString())));
            logger.error('\n\n');
            await help(config, mri(rest));
          } else {
            logger.error(error.toString());
          }
          if (process.env.NODE_ENV !== 'test') {
            process.exit(error.exitCode);
          }
        } else {
          logger.error(error);
        }
        if (process.env.NODE_ENV !== 'test') {
          process.exit(1);
        }
      })
      .catch(error => {
        logger.error(error.toString());
      });
  };
}

export default args;
