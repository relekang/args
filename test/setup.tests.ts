import { setup } from "../src/setup";

const defaultConfig = { name: 'cli', commands: {} };

test('setup should not do anything when needsSetup is not present', () => {
  setup(defaultConfig);
});

test('setup should not call setup when needsSetup resolves false', async () => {
  const config = {
    ...defaultConfig,
    needsSetup: async () => false,
    setup: jest.fn(),
  };

  await setup(config);

  expect(config.setup).not.toHaveBeenCalled();
});

test('setup should call setup when needsSetup resolves true', async () => {
  const config = {
    ...defaultConfig,
    needsSetup: async () => true,
    setup: jest.fn(),
  };

  await setup(config);

  expect(config.setup).toHaveBeenCalled();
});

test('setup should throw when needs setup and setup is missing', async () => {
  const config = {
    ...defaultConfig,
    needsSetup: async () => true,
  };

  await expect(setup(config)).rejects.toThrowErrorMatchingSnapshot();
});
