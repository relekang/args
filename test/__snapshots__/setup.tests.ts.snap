// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setup should throw when needs setup and setup is missing 1`] = `"This cli requires setup, however the setup is not provided."`;
